llama-rs/src/lib.rs

```diff
@@ -103,6 +105,9 @@ pub struct InferenceSession {
 
     /// The logits that were last predicted by the network. Zeroed out otherwise.
     last_logits: Vec<f32>,
+
+    /// Scratch buffers
+    scratch: [ggml::Buffer; 2]
```

Comment:
Do we know why it needs two buffers?

Reply:
> Do we know why it needs two buffers?

It's from llama.cpp. Same with the magic buffer sizes.

Reply:
Indeed. I don't know. And I'd _guess_ 7B wouldn't need as much, but I don't know and I didn't try, and even if I tried, I couldn't know what's the actual correct amount.

https://github.com/ggerganov/llama.cpp/blob/62cfc54f77e519057110265b52b0d614fa363e2a/llama.cpp#L63

Reply:
As for why it's two buffers, it's quite unclear to me. I'd guess ggml expects certain objects to live in scratch 2 while pointing to the scratch 1. I can say for sure it crashes if there's only one scratch buffer used (replace all `use_scratch` calls with the same buffer).

Reply:
And for the evaluate context memory requirements:
- https://github.com/ggerganov/llama.cpp/blob/62cfc54f77e519057110265b52b0d614fa363e2a/llama.cpp#L88
- https://github.com/ggerganov/llama.cpp/blob/62cfc54f77e519057110265b52b0d614fa363e2a/llama.cpp#L443

It's not really the greatest engineering to just copy the logic from llama.cpp without understanding it fully. At the moment I don't have better ideas though. I can try to think cleaner ways to do this context memory handling maybe next week, but that could be another PR as far as I'm concerned.

Reply:
Hm, OK, fair enough. We should document that we don't actually know the reasoning behind this, so it's clear that it's cargo-culting instead of some principled engineering choice.